File: purecoincidence/src/main/scala/purecoincidence/PurelyLong.scala
package purecoincidence

object PurelyLong extends App {
  def pure(a: Int, b: Int): Int =
    2*a*a + 3*b    // 2a^2 + 3b

  def expressionA(a: Int, b: Int): Int =
    4*a*a + pure(a, b) + 6*b

  def expressionB(a: Int, b: Int): Int =
    3 * pure(a, b)

  println((
    expressionA(3, 4),
    expressionB(3, 4)
  ))    // => (90,90)
}

File: hopefullypure/app/src/main/java/hopefullypure/PurelyLong.java
package hopefullypure;

import static java.lang.System.out;
import static java.lang.String.format;

public class PurelyLong {
    static int pure(int a, int b) {
        return 2 * a * a + 3 * b;
    }

    static int expressionA(int a, int b) {
        return 4 * a * a + pure(a, b) + 6 * b;
    }

    static int expressionB(int a, int b) {
        return 3 * pure(a, b);
    }

    public static void main(String[] args) {
        out.println(
          format("%d, %d",
            expressionA(3, 4),
            expressionB(3, 4)));
    }
}

